internal/docstore/firedocstore/query.go

```diff
@@ -66,18 +80,112 @@ func (it *docIterator) Next(ctx context.Context, doc driver.Document) error {
 		if err != nil {
 			return err
 		}
-		if res.Document != nil {
+		// No document => partial progress; keep receiving.
+		if res.Document == nil {
+			continue
+		}
+		match, err := it.evaluateLocalFilters(res.Document)
+		if err != nil {
+			return err
+		}
+		if match {
 			break
 		}
-
-		// No document => partial progress; keep receiving.
 	}
 	return decodeDoc(res.Document, doc, it.nameField)
 }
 
+// Report whether the filters are true of the document.
+func (it *docIterator) evaluateLocalFilters(pdoc *pb.Document) (bool, error) {
+	if len(it.localFilters) == 0 {
+		return true, nil
+	}
+	// TODO(jba): optimization: evaluate the filter directly on the proto document, without decoding.
+	m := map[string]interface{}{}
+	doc, err := driver.NewDocument(m)
+	if err != nil {
+		return false, err
+	}
+	if err := decodeDoc(pdoc, doc, it.nameField); err != nil {
+		return false, err
+	}
+	for _, f := range it.localFilters {
+		if !evaluateFilter(f, doc) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func evaluateFilter(f driver.Filter, doc driver.Document) bool {
+	val, err := doc.Get(f.FieldPath)
+	if err != nil {
+		// Treat a missing field as false.
+		return false
+	}
+	lhs := reflect.ValueOf(val)
+	rhs := reflect.ValueOf(f.Value)
+	if lhs.Kind() == reflect.String {
+		if rhs.Kind() != reflect.String {
+			return false
+		}
+		return applyComparison(f.Op, strings.Compare(lhs.String(), rhs.String()))
+	}
+	// Compare numbers by using big.Float. This is expensive
```

Comment:
Curious, does the firestore backend handle this the same way?

Reply:
I doubt it; it's expensive. (And I don't think the backend is written in Go.) It may use some larger floating-point type that can accurately represent any int64.